engine/audiomixer.cpp

```diff
@@ -0,0 +1,167 @@
+                  unsigned long framesPerBuffer,
+                  const PaStreamCallbackTimeInfo* timeInfo,
+                  PaStreamCallbackFlags statusFlags,
+                  void* userData)
+{
+    return Pong::GetInstance().GetAudioMixer().AudioCallback(inputBuffer, outputBuffer, framesPerBuffer, timeInfo, statusFlags, userData);
+}
+
+bool AudioMixer::Init()
+{
+    mVolume = Config::GetValue("volume", 0.5f);
+    LogInfo("Volume: {}", mVolume);
+
+    mSpatialAudioEnabled = Config::GetValue("spatial_audio", false);
+    LogInfo("Spatial Audio: {}", mSpatialAudioEnabled ? "Enabled" : "Disabled");
+
+    PaError err = Pa_Initialize();
+    if (err != paNoError) {
+        LogError("PortAudio initialization failed: {}", Pa_GetErrorText(err));
+    }
+
+    PaStreamParameters outputParameters;
+    outputParameters.device = Pa_GetDefaultOutputDevice();
+    outputParameters.channelCount = 2; // Stereo
+    outputParameters.sampleFormat = paFloat32;
+    outputParameters.suggestedLatency = Pa_GetDeviceInfo(outputParameters.device)->defaultLowOutputLatency;
+    outputParameters.hostApiSpecificStreamInfo = nullptr;
+
+    err = Pa_OpenStream(&mStream, nullptr, &outputParameters, SAMPLE_RATE, FRAMES_PER_BUFFER, paClipOff, AudioCallbackWrapper, nullptr);
+    if (err != paNoError) {
+        LogError("PortAudio stream opening failed: {}", Pa_GetErrorText(err));
+        Pa_Terminate();
+        return false;
+    }
+
+    err = Pa_StartStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream starting failed: {}", Pa_GetErrorText(err));
+        Pa_CloseStream(mStream);
+        Pa_Terminate();
+        return false;
+    }
+
+    return true;
+}
+
+void AudioMixer::Cleanup()
+{
+    PaError err = Pa_StopStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream stopping failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    err = Pa_CloseStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream closing failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    Pa_Terminate();
+}
+
+void AudioMixer::PlaySound(const Sound& sound)
+{
+    mPlayingSounds.emplace_back(sound);
+}
+
+void AudioMixer::PlaySound(const Sound& sound, const glm::vec3& position)
+{
+    mPlayingSounds.emplace_back(sound, position);
+}
+
+int AudioMixer::AudioCallback(const void* /*inputBuffer*/, void* outputBuffer,
+                              unsigned long framesPerBuffer,
+                              const PaStreamCallbackTimeInfo* /*timeInfo*/,
+                              PaStreamCallbackFlags /*statusFlags*/,
+                              void* /*userData*/)
+{
+    float* out = static_cast<float*>(outputBuffer);
+
+    int frames = static_cast<int>(framesPerBuffer);
+    for (int i = 0; i < frames; i++)
+    {
+        float sampleL = 0.0f;
+        float sampleR = 0.0f;
+
+        for (auto& sound : mPlayingSounds)
```

Comment:
Definitely a data race with these sounds at the moment. This would be a good question for Dave on how to handle this properly.  This is how easy it is to get into very nasty situations with threading and realtime safety, as they often conflict.

My initial thought on this is to use a a try_lock, but this is going to get quite complicated quickly. To not bog you down in this detail, I would recommend possibly just using a plain old mutex for now, and file an issue to deal with this later. 

This is how easy it is to fall of the "realtime safe" wagon :) 


The solution of how to do this properly likely exists in this talk:
https://www.youtube.com/watch?v=Q0vrQFyAdWI

Reply:
Yeah Ill just put a lock around these for now.